common/streams/src/sources/source_ndjson.rs

```diff
@@ -0,0 +1,156 @@
+// Copyright 2021 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::io::BufRead;
+
+use async_trait::async_trait;
+use common_datablocks::DataBlock;
+use common_datavalues::DataSchemaRef;
+use common_exception::ErrorCode;
+use common_exception::Result;
+use common_exception::ToErrorCode;
+
+use crate::Source;
+
+#[derive(Debug, Clone)]
+pub struct NDJsonSourceBuilder {
+    schema: DataSchemaRef,
+    block_size: usize,
+    size_limit: usize,
+}
+
+impl NDJsonSourceBuilder {
+    pub fn create(schema: DataSchemaRef) -> Self {
+        NDJsonSourceBuilder {
+            schema,
+            block_size: 10000,
+            size_limit: usize::MAX,
+        }
+    }
+
+    pub fn block_size(&mut self, block_size: usize) -> &mut Self {
+        self.block_size = block_size;
+        self
+    }
+
+    pub fn size_limit(&mut self, size_limit: usize) -> &mut Self {
+        self.size_limit = size_limit;
+        self
+    }
+
+    pub fn build<R>(&self, reader: R) -> Result<NDJsonSource<R>>
+    where R: BufRead + Unpin + Send {
+        NDJsonSource::try_create(self.clone(), reader)
+    }
+}
+
+pub struct NDJsonSource<R> {
+    builder: NDJsonSourceBuilder,
+    reader: R,
+    rows: usize,
+    buffer: String,
+}
+
+impl<R> NDJsonSource<R>
+where R: BufRead + Unpin + Send
+{
+    fn try_create(builder: NDJsonSourceBuilder, reader: R) -> Result<Self> {
+        Ok(Self {
+            builder,
+            reader,
+            rows: 0,
+            buffer: String::new(),
+        })
+    }
+}
+
+#[async_trait]
+impl<R> Source for NDJsonSource<R>
+where R: BufRead + Unpin + Send
+{
+    async fn read(&mut self) -> Result<Option<DataBlock>> {
+        // Check size_limit.
+        if self.rows >= self.builder.size_limit {
+            return Ok(None);
+        }
+
+        let mut packs = self
+            .builder
+            .schema
+            .fields()
+            .iter()
+            .map(|f| f.data_type().create_deserializer(self.builder.block_size))
+            .collect::<Vec<_>>();
+
+        let names = self
+            .builder
+            .schema
+            .fields()
+            .iter()
+            .map(|f| f.name())
+            .collect::<Vec<_>>();
+
+        let mut rows = 0;
+
+        loop {
+            self.buffer.clear();
+            if self
+                .reader
+                .read_line(&mut self.buffer)
```

Comment:
Here is a block IO in an async function, might block runtime thread, is this by design?

Reply:
Thanks, good catch, I will create a fix for that.